File: backend/src/user.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from './entities/user.entity';
import { StudentService } from './services/student.service';
import * as bcrypt from 'bcrypt';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    private studentService: StudentService,
  ) {}

  async createUser(username: string, password: string, imageUrl: string | null = null, studentId: number | null = null): Promise<void> {
    const passwordHash = await bcrypt.hash(password, 10);
    const user = this.userRepository.create({
      username,
      passwordHash,
      imageUrl,
      studentId,
      approvalStatus: 'approved', // کاربران موجود تایید شده هستند
      userType: 'student',
    });
    await this.userRepository.save(user);
  }

  async createPendingUser(
    username: string, 
    password: string, 
    imageUrl: string | null = null, 
    studentData: { firstName: string; lastName: string; email: string; phoneNumber: string }
  ): Promise<void> {
    const passwordHash = await bcrypt.hash(password, 10);
    const user = this.userRepository.create({
      username,
      passwordHash,
      imageUrl,
      approvalStatus: 'pending',
      userType: 'student',
    });
    await this.userRepository.save(user);
  }

  async getPendingUsers(): Promise<User[]> {
    return this.userRepository.find({
      where: { approvalStatus: 'pending' },
      relations: ['student']
    });
  }

  async approveUser(userId: number, studentData: { firstName: string; lastName: string; email: string; phoneNumber: string; studentId: string }): Promise<void> {
    // ابتدا دانش‌آموز را ایجاد کن
    const student = await this.studentService.create({
      ...studentData,
      status: 'active',
    });
    
    // سپس کاربر را تایید کن و به دانش‌آموز مرتبط کن
    await this.userRepository.update(userId, {
      approvalStatus: 'approved',
      studentId: student.id
    });
  }

  async rejectUser(userId: number): Promise<void> {
    await this.userRepository.update(userId, { approvalStatus: 'rejected' });
  }

  async linkUserToStudent(username: string, studentId: number): Promise<void> {
    await this.userRepository.update({ username }, { studentId });
  }

  async findUser(username: string): Promise<User | undefined> {
    return this.userRepository.findOne({ 
      where: { username },
      relations: ['student']
    });
  }

  async findUserByStudentId(studentId: number): Promise<User | undefined> {
    return this.userRepository.findOne({ 
      where: { studentId },
      relations: ['student']
    });
  }

  async findUserWithStudent(username: string): Promise<User | undefined> {
    return this.userRepository.findOne({ 
      where: { username },
      relations: ['student']
    });
  }

  async validateUser(username: string, password: string): Promise<boolean> {
    const user = await this.findUser(username);
    if (!user) {
      return false;
    }
    if (!user.passwordHash) {
      console.error('validateUser: passwordHash is null or undefined for user:', username);
      return false;
    }
    return bcrypt.compare(password, user.passwordHash);
  }

    async updateUserProfile(username: string, imageUrl: string): Promise<void> {
        await this.userRepository.update({ username }, { imageUrl });
    }
}

File: backend-dotnet/LessonPlanner.Api/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using LessonPlanner.Api.Data;
using LessonPlanner.Api.Models;

namespace LessonPlanner.Api.Services;

public class UserService : IUserService
{
    private readonly AppDbContext _db;

    public UserService(AppDbContext db)
    {
        _db = db;
    }

    public async Task CreateUserAsync(string username, string password, string? imageUrl, int? studentId, string userType, string? firstName = null, string? lastName = null, string? email = null, string? phoneNumber = null)
    {
        var hash = BCrypt.Net.BCrypt.HashPassword(password);
        var user = new User
        {
            Username = username,
            PasswordHash = hash,
            FirstName = firstName,
            LastName = lastName,
            Email = email,
            PhoneNumber = phoneNumber,
            ImageUrl = imageUrl,
            StudentId = studentId,
            ApprovalStatus = "approved",
            UserType = userType,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
        _db.Users.Add(user);
        await _db.SaveChangesAsync();
    }

    public async Task CreatePendingUserAsync(string username, string password, string? imageUrl, string? firstName = null, string? lastName = null, string? email = null, string? phoneNumber = null)
    {
        var hash = BCrypt.Net.BCrypt.HashPassword(password);
        var user = new User
        {
            Username = username,
            PasswordHash = hash,
            FirstName = firstName,
            LastName = lastName,
            Email = email,
            PhoneNumber = phoneNumber,
            ImageUrl = imageUrl,
            ApprovalStatus = "pending",
            UserType = "student",
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
        _db.Users.Add(user);
        await _db.SaveChangesAsync();
    }

    public async Task<List<User>> GetPendingUsersAsync()
    {
        return await _db.Users
            .Where(u => u.ApprovalStatus == "pending")
            .Include(u => u.Student)
            .ToListAsync();
    }

    public async Task ApproveUserAsync(int userId, string firstName, string lastName, string email, string phoneNumber, string studentId)
    {
        var student = new Student
        {
            FirstName = firstName,
            LastName = lastName,
            Email = email,
            PhoneNumber = phoneNumber,
            StudentId = studentId,
            Status = "active",
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };
        _db.Students.Add(student);
        await _db.SaveChangesAsync();

        var user = await _db.Users.FindAsync(userId);
        if (user != null)
        {
            user.ApprovalStatus = "approved";
            user.StudentId = student.Id;
            user.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync();
        }
    }

    public async Task RejectUserAsync(int userId)
    {
        var user = await _db.Users.FindAsync(userId);
        if (user != null)
        {
            user.ApprovalStatus = "rejected";
            user.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync();
        }
    }

    public async Task<User?> FindUserAsync(string username)
    {
        return await _db.Users
            .Include(u => u.Student)
            .FirstOrDefaultAsync(u => u.Username == username);
    }

    public async Task<User?> FindUserByStudentIdAsync(int studentId)
    {
        return await _db.Users
            .Include(u => u.Student)
            .FirstOrDefaultAsync(u => u.StudentId == studentId);
    }

    public async Task<bool> ValidateUserAsync(string username, string password)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Username == username);
        if (user?.PasswordHash == null) return false;
        return BCrypt.Net.BCrypt.Verify(password, user.PasswordHash);
    }

    public async Task UpdateUserProfileAsync(string username, string imageUrl)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Username == username);
        if (user != null)
        {
            user.ImageUrl = imageUrl;
            user.UpdatedAt = DateTime.UtcNow;
            await _db.SaveChangesAsync();
        }
    }
}
